File: Sources/MacKVMCore/InboundPayloadBudget.swift
import Foundation

/// A deterministic per-second packet and byte budget for authenticated input.
/// It is intentionally independent of the network stack so boundary behavior
/// can be tested without opening a socket.
public struct InboundPayloadBudget: Equatable, Sendable {
    public let maximumPacketsPerSecond: Int
    public let maximumBytesPerSecond: Int

    private var windowStart: UInt64 = 0
    private var windowBytes = 0
    private var windowPackets = 0

    public init(
        maximumPacketsPerSecond: Int,
        maximumBytesPerSecond: Int
    ) {
        self.maximumPacketsPerSecond = maximumPacketsPerSecond
        self.maximumBytesPerSecond = maximumBytesPerSecond
    }

    public mutating func allows(bytes: Int, at now: UInt64) -> Bool {
        guard maximumPacketsPerSecond > 0,
              maximumBytesPerSecond > 0,
              bytes >= 0,
              bytes <= maximumBytesPerSecond else {
            return false
        }

        if windowStart == 0
            || now < windowStart
            || now - windowStart >= 1_000_000_000 {
            windowStart = now
            windowBytes = 0
            windowPackets = 0
        }

        guard windowPackets < maximumPacketsPerSecond,
              windowBytes <= maximumBytesPerSecond - bytes else {
            return false
        }
        windowPackets += 1
        windowBytes += bytes
        return true
    }

    public mutating func reset() {
        windowStart = 0
        windowBytes = 0
        windowPackets = 0
    }
}

File: WindowsKVM/src/WindowsKVM.Protocol/InboundPayloadBudget.cs
namespace WindowsKVM.Protocol;

/// <summary>
/// Bounds authenticated input traffic per one-second window. The budget is
/// deliberately a rolling window rather than a lifetime packet count: a
/// normal mouse or trackpad stream must remain connected while an abusive
/// peer still cannot consume unbounded CPU or memory.
/// </summary>
public sealed class InboundPayloadBudget
{
    private long windowStartMilliseconds = -1;
    private int windowPackets;
    private long windowBytes;

    public InboundPayloadBudget(int maximumPacketsPerSecond, long maximumBytesPerSecond)
    {
        if (maximumPacketsPerSecond <= 0 || maximumBytesPerSecond <= 0)
        {
            throw new ArgumentOutOfRangeException();
        }

        MaximumPacketsPerSecond = maximumPacketsPerSecond;
        MaximumBytesPerSecond = maximumBytesPerSecond;
    }

    public int MaximumPacketsPerSecond { get; }

    public long MaximumBytesPerSecond { get; }

    public bool Allows(int bytes, long nowMilliseconds)
    {
        if (bytes < 0 || bytes > MaximumBytesPerSecond || nowMilliseconds < 0)
        {
            return false;
        }

        if (windowStartMilliseconds < 0
            || nowMilliseconds < windowStartMilliseconds
            || nowMilliseconds - windowStartMilliseconds >= 1_000)
        {
            windowStartMilliseconds = nowMilliseconds;
            windowPackets = 0;
            windowBytes = 0;
        }

        if (windowPackets >= MaximumPacketsPerSecond
            || windowBytes > MaximumBytesPerSecond - bytes)
        {
            return false;
        }

        windowPackets++;
        windowBytes += bytes;
        return true;
    }

    public void Reset()
    {
        windowStartMilliseconds = -1;
        windowPackets = 0;
        windowBytes = 0;
    }
}
